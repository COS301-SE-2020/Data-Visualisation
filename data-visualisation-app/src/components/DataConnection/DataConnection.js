/**
 *   @file DataConnection.js
 *   Project: Data Visualisation Generator
 *   Copyright: Open Source
 *   Organisation: Doofenshmirtz Evil Incorporated
 *
 *   Update History:
 *   Date        Author              Changes
 *   -------------------------------------------------------
 *   8/7/2020   Byron Tominson      Original
 *   15/7/2020  Byron Tominson      API calls added
 *
 *   Test Cases: data-visualisation-app/src/tests/DataConnection.test.js
 *
 *   Functional Description:
 *   Displays data connections to the user.
 *
 *   Error Messages: "Error"
 *   Assumptions: None
 *   Constraints: None
 */

/**
  * Imports
*/ 
import React from 'react';
import { List, Avatar, Button, Skeleton } from 'antd';
import request from '../../globals/requests';
import * as constants from '../../globals/constants';
import './DataConnection.scss';
import AddConnectionDialog from '../AddConnectionDialog';

/**
  * takes no arguments and returns a random string of length 10.
  * @return result
*/
export function generateID() {
  let result = '';
  let characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  let charactersLength = characters.length;
  for (let i = 0; i < 10; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
}


/**
  * a class comonent that renders the data connection list.
  * @returns React Component
*/
class DataConnection extends React.Component {

  
  /**
    * States that belong to the Data Connection class component.
  */
  state = {
    initLoading: true,
    loading: false,
    data: [],
    list: [],
    addConnection : false,
  };


  /**
    * invoked immediately after a component is mounted (inserted into the tree).
  */
  componentDidMount() {
    console.log(request.user.isLoggedIn);
      this.getData(res => {
        this.setState({
          initLoading: false,
          data: request.user.dataSources,
          list: request.user.dataSources,
        });
      });
  }
  
  /**
    * Calls the function to send the request to get the list of data sources.
    * The function called updates the 'requests.user.dataSources' object with the correct array of data sources for the user. 
  */
  getData = callback => {
    if(request.user.isLoggedIn){
      request.dataSources.list(request.user.apikey, function(result) {
        console.log(result);
        
        if (result === constants.RESPONSE_CODES.SUCCESS) {
          callback(request.user.dataSources);
        }
      });
    }
    else{
      callback(request.user.dataSources);
    } 
  };

  /**
    * Deletes item from the 'list'.
    * If user is logged in, delete on backend and front end.
    * If the user is not logged in, delete only on front end.
  */
  deleteItem = (itemToDelete) => {
    /**
      * If user is logged in, delete on backend and front end.
    */
    if(request.user.isLoggedIn){
      
      request.dataSources.delete(itemToDelete.id, request.user.apikey, function(result) {
        console.log(result);
        if (result === constants.RESPONSE_CODES.SUCCESS) {

        }
      });

      request.user.dataSources = this.state.list.filter(list => list.id !== itemToDelete.id);

      this.setState(previousState => ({
        data: request.user.dataSources,
        list: request.user.dataSources
      }));
    }
    else{
      /**
        * User is not logged in, delete only on front end.
      */
      request.user.dataSources = this.state.list.filter(list => list.id !== itemToDelete.id);

      this.setState(previousState => ({
        data: request.user.dataSources,
        list: request.user.dataSources
      }));
    }
   
  }


  /**
    * Adds item to the 'list'.
    * If user is logged in, delete on backend and front end.
    * If user is not logged in, add item only front end.
  */
  addItem = (values) => {
    var id = generateID();
    let attempt = this;
    console.log(request.user.apikey);
    /**
      * If user is logged in, add item on backend and front end.
    */
    if(request.user.isLoggedIn){
      request.dataSources.add(request.user.apikey, values.uri, function(result) {
       
        if (result === constants.RESPONSE_CODES.SUCCESS) {

          request.user.dataSources.push({
            'id': request.user.addedSourceID,
            'email': request.user.email,
            'sourceurl': values.uri
          });

          attempt.setState(previousState => ({
            data: request.user.dataSources,
            list: request.user.dataSources
          }));

        }
      });
      
    }
    else{
      /**
        * User is not logged in, add item only front end.
      */
      request.user.dataSources.push({
        'id': id,
        'email': request.user.email,
        'sourceurl': values.uri
      });

      this.setState(previousState => ({
        data: request.user.dataSources,
        list: request.user.dataSources
      }));
    }
    
  }

 
  /**
    * handles the addConnection state
  */
  changeAddState = () => {
    this.setState({
      addConnection : !this.state.addConnection,
    });
  };

  /**
    * sets the stage of explore to entities (the next stage)
  */
  next = () => {
    this.props.setStage('entities');
  };

  /**
    * renders elements
  */
  render() {
    const { initLoading, loading, list } = this.state;
    const loadMore =
      !initLoading && !loading ? (
        <div
          style={{
            textAlign: 'center',
            marginTop: 12,
            height: 32,
            lineHeight: '32px',
          }}
        >
          <Button shape = 'round' onClick={this.changeAddState} style ={{marginRight: '10px'}}>Add Connection</Button>
          <Button type = 'primary' shape = 'round' onClick={this.next}>Next</Button>
        </div>
      ) : null;
   

    return (
     <div>
        <List
          className="dataSourceList"
          loading={initLoading}
          itemLayout="horizontal"
          loadMore={loadMore}
          dataSource={list}
          renderItem={item => (
            <List.Item
              key={item.id}
              actions={
                [
                  <Button id = 'deleteButton' onClick={() => {this.deleteItem(item);}} >delete</Button>
                ]
              }>
              <Skeleton avatar title={false} loading={item.loading} active>
                <List.Item.Meta
                  avatar={
                    <Avatar src="https://15f76u3xxy662wdat72j3l53-wpengine.netdna-ssl.com/wp-content/uploads/2018/03/OData-connector-e1530608193386.png" />
                  }
                  title={item.sourceurl}
                  description={item.id}
                />
                <div></div>
              </Skeleton>
            </List.Item>
          )}
        />
        <main>
        {
            this.state.addConnection ? 
            
              <AddConnectionDialog changeState = {this.changeAddState} addItem = {this.addItem}/>
            
            :
            null
        }
        </main>
      </div>
    );
  }
}

export default DataConnection;