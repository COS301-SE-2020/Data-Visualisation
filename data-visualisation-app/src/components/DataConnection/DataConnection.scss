@import '../../globals/globals.scss';


.dataConnectionCard{ 

  border-radius: 5px;
  box-shadow:
    0 2.8px 2px rgba(0, 0, 0, 0.004),
    0 6.7px 5px rgba(0, 0, 0, 0.008),
    0 1px 1px rgba(0, 0, 0, 0.006),
    0 2.3px 17.9px rgba(0, 0, 0, 0.002),
    0 4.8px 3.4px rgba(0, 0, 0, 0.006),
    0 1px 8px rgba(0, 0, 0, 0.02);

  width: 80%;
  margin: 40px auto;
  background: white;
  border-radius: 5px;
  word-wrap: break-word;
}

#addConnetion__button{
  background-color: white;
  border: none;
  color: $color-primary;
  -webkit-box-shadow: 0px 2px 2px -2px $color-primary;
  -moz-box-shadow: 0px 2px 2px -2px $color-primary;
  box-shadow: 0px 2px 2px -2px $color-primary;
  font-weight: 700px;
}


#next__button{
    background-color: $color-primary;
    border: none;
    color: white;
    opacity: 0.9;
}

#next__button:hover{
  opacity: 1;
 
}

@media only screen and (max-width: 700px) {
  .dataConnectionCard{
      
    width: 95%;
  
  }
}

.connectionHeader{
  width:'80%'; 
  margin: '0 auto';
  text-align: center; 
  margin-top: 36px;
  font-size: 1.175em;
  font-weight: 500;
}

