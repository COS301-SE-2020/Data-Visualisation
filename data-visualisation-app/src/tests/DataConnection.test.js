import React from 'react';
import DataConnection, {generateID} from '../components/DataConnection/DataConnection';
import renderer from 'react-test-renderer';

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

describe('ID generator', () => {
    it('It reutrns the right type of string', () => {
      expect(generateID()).toHaveLength(10);
    });

});

describe('DataConnection component', () => {
  it('renders correctly', () => {
    const rendered = renderer.create(<DataConnection/>);
    expect(rendered.toJSON()).toMatchSnapshot();
  });
});

