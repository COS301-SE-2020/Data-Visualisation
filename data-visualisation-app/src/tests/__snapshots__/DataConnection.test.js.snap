// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataConnection component renders correctly 1`] = `
<div>
  <div
    className="ant-list dataSourceList ant-list-split ant-list-something-after-last-item"
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
      >
        <ul
          className="ant-list-items"
        >
          <li
            className="ant-list-item"
          >
            <div
              className="ant-list-item-meta"
            >
              <div
                className="ant-list-item-meta-avatar"
              >
                <span
                  className="ant-avatar ant-avatar-circle ant-avatar-image"
                  style={Object {}}
                >
                  <img
                    onError={[Function]}
                    src="https://15f76u3xxy662wdat72j3l53-wpengine.netdna-ssl.com/wp-content/uploads/2018/03/OData-connector-e1530608193386.png"
                  />
                </span>
              </div>
              <div
                className="ant-list-item-meta-content"
              >
                <h4
                  className="ant-list-item-meta-title"
                >
                  <a
                    href="https://ant.design"
                  >
                    https://services.odata.org/V2/Northwind/Northwind.svc
                  </a>
                </h4>
                <div
                  className="ant-list-item-meta-description"
                >
                  6
                </div>
              </div>
            </div>
            <div />
            <ul
              className="ant-list-item-action"
            >
              <li>
                <button
                  className="ant-btn"
                  id="deleteButton"
                  onClick={[Function]}
                  type="button"
                >
                  <span>
                    delete
                  </span>
                </button>
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </div>
    <div
      style={
        Object {
          "height": 32,
          "lineHeight": "32px",
          "marginTop": 12,
          "textAlign": "center",
        }
      }
    >
      <button
        className="ant-btn ant-btn-round"
        onClick={[Function]}
        style={
          Object {
            "marginRight": "10px",
          }
        }
        type="button"
      >
        <span>
          Add Connection
        </span>
      </button>
      <button
        className="ant-btn ant-btn-primary ant-btn-round"
        onClick={[Function]}
        type="button"
      >
        <span>
          Next
        </span>
      </button>
    </div>
  </div>
  <main />
</div>
`;
